File: docs/csharp/scripts/async.cs
#:sdk Microsoft.NET.Sdk
// Princípios: async/await, Task.WhenAll para operações independentes

var dashboard = await FetchDashboard(1);
Console.WriteLine($"user:   {dashboard.User}");
Console.WriteLine($"orders: {string.Join(", ", dashboard.Orders)}");

async Task<Dashboard> FetchDashboard(int userId)
{
    var userTask = FetchUser(userId);
    var ordersTask = FetchOrders(userId);

    await Task.WhenAll(userTask, ordersTask);

    var user = await userTask;
    var orders = await ordersTask;
    var dashboard = new Dashboard(user, orders);
    return dashboard;
}

async Task<User> FetchUser(int id)
{
    await Task.Delay(100);
    var user = new User(id, "Alice");
    return user;
}

async Task<IReadOnlyList<Order>> FetchOrders(int userId)
{
    await Task.Delay(80);
    var orders = new List<Order> { new(10, userId, 150m) };
    return orders;
}

record User(int Id, string Name);
record Order(int Id, int UserId, decimal Total);
record Dashboard(User User, IReadOnlyList<Order> Orders);

File: docs/typescript/scripts/async.ts
// Princípios: async/await, Promise.all, dependências explícitas, entry point limpo

type User = {
  readonly id: number;
  readonly name: string;
};

type Order = {
  readonly id: number;
  readonly userId: number;
  readonly total: number;
};

type Dashboard = {
  readonly user: User;
  readonly orders: Order[];
};

fetchDashboard(1).then(console.log);

async function fetchDashboard(userId: number): Promise<Dashboard> {
  const [user, orders] = await Promise.all([
    fetchUser(userId),
    fetchOrders(userId),
  ]);

  const dashboard: Dashboard = { user, orders };

  return dashboard;
}

async function fetchUser(id: number): Promise<User> {
  await wait(100);
  const user: User = { id, name: "Alice" };

  return user;
}

async function fetchOrders(userId: number): Promise<Order[]> {
  await wait(80);
  const orders: Order[] = [{ id: 10, userId, total: 150 }];

  return orders;
}

function wait(ms: number): Promise<void> {
  const delay = new Promise<void>((resolve) => setTimeout(resolve, ms));
  return delay;
}
